### DISCUSSION TITLE
Excluding path from rewrites

### DISCUSSION THREAD

**Discussion by jir-f** [vercel/next.js#40945] [Help] [answered]
Hey there!

I'm looking into excluding a few paths from a rewrite but it doesn't seem like I'm able to get it to work.
I currently have this in the rewrite but it doesn't seem like it is working.

`/blog/:slug(((?!post|guides-and-diy).*))/:uid`

I would like allow anything but exact matches to `post` and `guides-and-diy`. So `sample-post` will still be a valid rewrite.
I've tried testing the regex [here](https://regex101.com/r/X8edpj/1) but as you'll see it is not working as expected

Any suggestion on how to do this?

**Comment by icyJoseph** [comment, marked as answer]
Try with:

```js
 "/blog/:slug((?!post|guides-and-diy).*)/:uid",
```

## Edit

The above missed to reject cases where `post` was part of the string:

```js
const path = "/blog/:slug((?!post(?=/)|guides-and-diy(?=/)).*)/:uid";
```

This fixes it by taking advantage of look ahead, because the slug is bound by `/`.

### Test cases

```js
const { pathToRegexp } = require("path-to-regexp");

const path = "/blog/:slug((?!post(?=/)|guides-and-diy(?=/)).*)/:uid";

const regexp = pathToRegexp(path);

console.log(regexp.test("/blog/post/abc"), false);
console.log(regexp.test("/blog/guides-and-diy/abc"), false); 
console.log(regexp.test("/blog/sample-post/abc"), true); 
console.log(regexp.test("/blog/sample/abc"), true); 
console.log(regexp.test("/blog/posts/abc"), true); 
console.log(regexp.test("/blog/post-sample/abc"), true);
console.log(regexp.test("/blog/guides-and-diys/abc"), true); 
console.log(regexp.test("/blog/sample-post/abc"), true)
console.log(regexp.test("/blog/posts/abc"), true); 

```

**Reply by jir-f** [reply]
That doesn't seem to work. If you look at the regex tester [here](https://regex101.com/r/TyxQpA/1) you can see all the words have matches.
I've also tested it in the project and it's only working for `guides-and-diy` but not for `post`. It is rewriting if the slug is `sample-post` or just `post`.

**Reply by icyJoseph** [reply]
Ah perhaps I copied the wrong solution, because I tested this. Let me try again.

**Reply by icyJoseph** [reply]
Ah you are right, I missed one scenario:

```js
const { pathToRegexp } = require("path-to-regexp");

const path = "/blog/:slug((?!post|guides-and-diy).*)/:uid";

const regexp = pathToRegexp(path);

console.log(regexp.test("/blog/post/abc")); // false
console.log(regexp.test("/blog/guides-and-diy/abc")); //false
console.log(regexp.test("/blog/sample-post/abc")); // true
console.log(regexp.test("/blog/sample/abc")); // true
console.log(regexp.test("/blog/posts/abc")); // false
```

Back to the tweaking lab... you can also just play around with `path-to-regexp` instead of using other utilities.

Or wait, `posts` is ok for your example right?

**Reply by icyJoseph** [reply]
@jir-f 

I am confused now... Let's go through a checklist

- If I go to `/blog/posts/abc` stay there.
- If I go to `/blog/guides-and-diy/123` stay there.
- I I go to `/blog/sample-post/xyz` redirect to, for example, `/news/sample-post/xyz`
- If I go to `/blog/posts/foo` redirect to, `/news/posts/foo`

Is that what you want?

Using this `next.config.js`:

```js
/** @type {import('next').NextConfig} */
const nextConfig = {
  reactStrictMode: true,
  async redirects() {
    return [
      {
        source: "/blog/:slug((?!post|guides-and-diy).*)/:uid",
        destination: "/news/:slug/:uid",
        permanent: false,
      },
    ];
  },
};

module.exports = nextConfig;
```

**Reply by jir-f** [reply]
So basically I'm trying to get only the exact matches of the two slugs `post` and `guides-and-diy` and if the slug is `posts` it will rewrite the url.

- If I go to /blog/post/foo stay there
- If I go to /blog/guides-and-diy/123 stay there.
- If I go to /blog/posts/abc rewrite to /sample-site/posts/abc.
- If I go to /blog/sample-post/xyz redirect to, for example, /news/sample-post/xyz
- If I go to /blog/posts/foo redirect to, /news/posts/foo

The only thing that is extra from what you mentioned is to allow to rewrite the url with the slug `posts`

**Reply by icyJoseph** [reply]
Aha, I tried with both `rewrites` and `redirects` and it's all good, except the `/blog/posts/foo` case, where the `regex-like-path` stays, at `blog` I see now.  

Ugh, I hope this is not one of those cases where the negative look ahead is really playing us.

**Reply by jir-f** [reply]
Yeah everything is working fine except for that bit.

I've also tried it with these slugs and it is still not working as exact matching of the strings 😑

```js
const { pathToRegexp } = require("path-to-regexp");

const path = "/blog/:slug((?!post|guides-and-diy).*)/:uid";

const regexp = pathToRegexp(path);

console.log(regexp.test("/blog/post-sample/abc")); // false -> should be true
console.log(regexp.test("/blog/guides-and-diys/abc")); // false -> should be true
console.log(regexp.test("/blog/sample-post/abc")); // true
console.log(regexp.test("/blog/posts/abc")); // false -> should be true

```

**Reply by icyJoseph** [reply]
Try this one please: `const path = "/blog/:slug((?!post(?=/)|guides-and-diy(?=/)).*)/:uid";`

**Reply by jir-f** [reply]
Yup that fixed the problem! I didn't think of adding `(?=/)`. Thank you so much for your help 🎉  @icyJoseph

**Reply by icyJoseph** [reply]
I'll update the main answer! Yeah look ahead felt like a good way out here, since the slug is bound by `/` - not sure it is super generic, but it gets the job done 😅 - you also learned to use `path-to-regexp`, so you could even write tests for this matchers. Anyway, please don't forget to mark as answered.
